### PR TITLE
pcsx2: 1.7.5587 -> 1.7.5779

### PR DESCRIPTION
**Pull request by SuperSamus**
## Description of changes

Diff: https://github.com/PCSX2/pcsx2/compare/v1.7.5587...v1.7.5779

~Depends on #305508 (which is the base of this branch).~ Merged.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by SuperSamus** [comment]
Waiting to know whether it's better to patch `shaderc` inside the package or in `all-packages`. (#309064)

**Comment by hrdinka** [comment]
@SuperSamus thanks a lot for continuously updating this package! As always I am very happy about your contribution.

Regarding the `shaderc` patch: IMHO it should be overriden in `all-packages.nix`. There are many other packages doing it this way as well. The `pcsx2` derivation should be agnostic of its input and not try to override them, it cannot know what exact derivations are passed as input. It should basically just ask for a `shaderc` package and be happy with whatever it gets. The caller should make sure to pass the right one. If a package needs a special version of an input, it is always done via `callPackage` in `all-packages.nix` and this here is not different.

Further a `fetch*` variant should be preferred and used whenever possible like in the duckstation PR to not bloat nixpkgs unnecessarily.

I would however not be happy about moving it to the bottom in `all-packages.nix` like it was proposed in the duckstation PR. I appreciate its good intentions but in the end this will mess up the ordering in `all-packages.nix` even more since I expect only few people to notice and adhere to it. In the light of RCF140, eventually all default package calls will be removed from `all-packages.nix` anyway.

File: pkgs/by-name/pc/pcsx2/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/pc/pcsx2/package.nix:45]: qt6Packages.callPackages should be used instead of inheriting here, like it was done before.

**Review by SuperSandro2000 — DISMISSED** [review]

File: pkgs/by-name/pc/pcsx2/package.nix
Comment (AndersonTorres) [reply pkgs/by-name/pc/pcsx2/package.nix]: Why?

**Comment by AndersonTorres** [comment]
```
Run if result/bin/nixpkgs-check-by-name --base "$base" .; then
- Because pkgs/by-name/pc/pcsx2 exists, the attribute `pkgs.pcsx2` must be defined like

    pcsx2 = callPackage ./../by-name/pc/pcsx2/package.nix { /* ... */ };

  However, in this PR, a different `callPackage` is used. See the definition in pkgs/top-level/all-packages.nix:2752:

    pcsx2 = qt6Packages.callPackage ../by-name/pc/pcsx2/package.nix {
      shaderc = shaderc.overrideAttrs (old: {
        patches = (old.patches or [ ]) ++ [
          (fetchpatch {
            url = "file://${pcsx2.src}/.github/workflows/scripts/common/shaderc-changes.patch";
            hash = "sha256-cD/gZwo7ARoePbEsHGHV34sx6MomfNOM7BSZDH1NlZM=";
            excludes = [
              "CHANGES"
              "CMakeLists.txt"
              "libshaderc/CMakeLists.txt"
            ];
          })
        ];
      });
    };
```

File: pkgs/top-level/all-packages.nix
Comment (hrdinka) [inline pkgs/top-level/all-packages.nix:2763]: Hash should be `sha256-Ps/D+CdSbjVWg3ZGOEcgbpQbCNkI5Nuizm4E5qiM9Wo=`

**Review by hrdinka — CHANGES_REQUESTED** [review]

File: pkgs/by-name/pc/pcsx2/package.nix
Comment (SuperSandro2000) [reply pkgs/by-name/pc/pcsx2/package.nix:45]: Same reason as in the other PR. pkgs-by-name is not meant to replace existing wrappers with more copy paste code and for those it is totally fine to keep the current way.

File: pkgs/by-name/pc/pcsx2/package.nix
Comment (SuperSandro2000) [inline pkgs/by-name/pc/pcsx2/package.nix:113]:
```suggestion
  qtWrapperArgs = [
    "--prefix LD_LIBRARY_PATH : ${lib.makeLibraryPath ([ vulkan-loader ] ++ cubeb.passthru.backendLibs)}"
  ];
```

File: pkgs/top-level/all-packages.nix
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:2772]: IMO we should move that inside the package. Otherwise it is easily missed and forgotten.

Comment (hrdinka) [reply pkgs/top-level/all-packages.nix:2772]: I agree with you but I don’t know ho to do this better? What is your solution to this?

When a custom dependency is needed, it should be passed just like this in `all-packages.nix`. This is how it is done for any derivation that needs different inputs. The derivation itself just uses whatever it transparently gets handed to. Of course here is more inline code due the patching, but there are also other packages doing it like this.

Patching an transparently passed dependency within the derivation (as initially proposed) is not a good solution (see my previous comment).

We could add a new non-existing input to the package with a default value (e.g. `shaderc-patched`). This way it is in pcsx2 package file and it could still be overridden by users. However if a new derivation is added named the same it will break. Further this breaks with the convention of how it is done with every other derivation.

So overall this seems to be the cleanest and best approach to me but I would be happy about suggestions.

**Comment by hrdinka** [comment]
@SuperSamus thanks a lot for your work so far! Currently merging is blocked since there still are change requests that need to be addressed.

**Comment by SuperSamus** [comment]
I'm waiting on what they decide to do on  #309064 (which has the same exact problem).

**Comment by hrdinka** [comment]
Ok, all clear

File: pkgs/top-level/all-packages.nix
Comment (SuperSandro2000) [reply pkgs/top-level/all-packages.nix:2772]: > When a custom dependency is needed, it should be passed just like this in `all-packages.nix`. This is how it is done for any derivation that needs different inputs. The derivation itself just uses whatever it transparently gets handed to. Of course here is more inline code due the patching, but there are also other packages doing it like this.

It can also be placed in a let in.

Comment (hrdinka) [reply pkgs/top-level/all-packages.nix:2772]: Yes, but the problem is that the possibility of overriding `shaderc` is either lost or it does unexpected stuff with your own supplied `shaderc` package.

Initially (in this PR) the derivation got `shaderc` passed via an argument and patched it within the derivation. By doing so `shaderc` was theoretically overridable. It would however mess with the package supplied by the user and could fail in a patch conflict. Thus rendering this a bad solution because it does things where they should not be done leading to unexpected behavior.

In the duckstation PR `shaderc` is not passed as an argument (only indirectly via `callPackage`). So it is not overridable as usually, which is not great either.

The current solution makes the derivation behave as it should, without any unexpected behavior or caveats. Only problem is the patch that could easily be overlooked. So I suggest to fix this exact problem by simply putting a comment besides the `shaderc` argument that reminds about this fact.

Comment (AndersonTorres) [reply pkgs/top-level/all-packages.nix:2772]: My rationale for duckstation was that the Duckstation developers are treating shaderc as a vendored dependency.
The difference is that they are not providing the whole Git subproject themselves, but a patch.

Then we should basically create a new package, `shaderc-specifically-patched-for-duckstation`.
But since this package is not meant to be useful outside duckstation, it is not exposed to top-level.

**Comment by SuperSamus** [comment]
Rebased, done what was done with DuckStation, and backported upstream patches to fix startup crashes

(Current master requires some stuff that nixpkgs doesn't have yet, like an updated SDL2 and Shaderc. One step at a time, this PR has been open long enough already.)

File: pkgs/by-name/pc/pcsx2/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/pc/pcsx2/package.nix:146]:
```suggestion
    license = with licenses; [ gpl3Plus lgpl3Plus ];
```

File: pkgs/by-name/pc/pcsx2/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/pc/pcsx2/package.nix:73]:
```suggestion
    (lib.cmakeBool "DISABLE_ADVANCE_SIMD"  true)
    (lib.cmakeBool "USE_LINKED_FFMPEG"  true)
```

File: pkgs/by-name/pc/pcsx2/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/pc/pcsx2/package.nix:74]:
```suggestion
    (lib.cmakeFeature "PCSX2_GIT_REV" finalAttrs.src.rev)
```

**Review by AndersonTorres — COMMENTED** [review]
Some minor cleanup.

File: pkgs/by-name/pc/pcsx2/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/pc/pcsx2/package.nix:61]: Oh man, I did not see these!

Please comment why each patch is needed here.

**Review by AndersonTorres — APPROVED** [review]
Besides, LGTM

**Review by AndersonTorres — APPROVED** [review]

**Comment by AndersonTorres** [comment]
P.S.: we should have a team for emulators.

**Comment by SuperSamus** [comment]
Anything left to do before this can be merged?

File: pkgs/by-name/pc/pcsx2/package.nix
Comment (AndersonTorres) [inline pkgs/by-name/pc/pcsx2/package.nix:150]: Just per curiousity,  what about setting this as Linux?

**Review by hrdinka — APPROVED** [review]

**Comment by hrdinka** [comment]
Tested locally. Thanks to everyone and thanks @SuperSamus for maintaining pcsx2 🙂

I had to dismiss the outdated review which blocked the merge.

**Pull request merged**
